ludwig/benchmarking/reporting.py

```diff
@@ -1,35 +1,166 @@
+from torch.autograd import DeviceType, profiler_util
+
+from ludwig.constants import LUDWIG_TAG
+
+
+def initialize_stats_dict(main_function_events: List[profiler_util.FunctionEvent]) -> Dict[str, Any]:
+    """Initialize dictionary which stores resource usage information per tagged code block.
+
+    :param main_function_events: list of main function events.
+    """
+    info = dict()
+    for event_name in [evt.name for evt in main_function_events]:
+        info[event_name] = []
+    return info
+
+
+def get_memory_details(kineto_event: _KinetoEvent) -> Tuple[str, int]:
+    """Get device name and number of bytes (de)allocated during an event.
+
+    :param kineto_event: a Kineto event instance.
+    """
+    if kineto_event.device_type() in [DeviceType.CPU, DeviceType.MKLDNN, DeviceType.IDEEP]:
+        return "cpu", kineto_event.nbytes()
+    elif kineto_event.device_type() in [DeviceType.CUDA, DeviceType.HIP]:
+        return f"cuda_{kineto_event.device_index()}", kineto_event.nbytes()
+    else:
+        raise ValueError(f"Device {kineto_event.device_type()} is not valid.")
+
+
+def get_device_memory_usage(
+    kineto_event: _KinetoEvent, mem_records_acc: profiler_util.MemRecordsAcc, run_usage_info: Dict[str, Any]
+) -> Dict[str, Any]:
+    """Get CPU and CUDA memory usage for an event.
+
+    :param kineto_event: a Kineto event instance.
+    :param mem_records_acc: memory events wrapped in a data structure for fast lookup.
+    :param run_usage_info: usage info for one execution of the tagged code block.
+    """
+    records_in_interval = mem_records_acc.in_interval(
+        kineto_event.start_us(), kineto_event.start_us() + kineto_event.duration_us()
+    )
+    memory_so_far = defaultdict(int)
+    count_so_far = defaultdict(int)
+    average_so_far = defaultdict(float)
+    max_so_far = defaultdict(int)
+
+    for mem_record in records_in_interval:
+        device, nbytes = get_memory_details(mem_record[0])
+        memory_so_far[device] += nbytes
+        max_so_far[device] = max(max_so_far[device], memory_so_far[device])
+        average_so_far[device] = (memory_so_far[device] + (average_so_far[device] * count_so_far[device])) / (
+            count_so_far[device] + 1
+        )
+        count_so_far[device] += 1
+    for device in count_so_far:
+        run_usage_info[f"torch_average_{device}_memory_usage"] = average_so_far[device]
+        run_usage_info[f"torch_max_{device}_memory_usage"] = max_so_far[device]
+    return run_usage_info
+
+
+def get_torch_op_time(events: List[profiler_util.FunctionEvent], attr: str) -> Union[int, float]:
+    """Get time torch operators spent executing for a list of events.
+
+    :param events: list of events.
+    :param attr: a FunctionEvent attribute. Expecting one of "cpu_time_total", "cuda_time_total".
+    """
+    if attr not in ["cpu_time_total", "cuda_time_total"]:
+        return -1
+
+    total = 0
+    for e in events:
+        if LUDWIG_TAG not in e.trace_name:
```

Comment:
Which events have LUDWIG_TAG in the trace name and which do not? Probably worth a comment.